src/plugins/data/common/data_frames/utils.ts

```diff
@@ -472,3 +472,77 @@ export const dataFrameToSpec = (dataFrame: IDataFrame, id?: string): IndexPatter
     fields: fields.reduce(flattenFields, {} as IndexPatternFieldMap),
   };
 };
+
+type FetchFunction<T, P = void> = (params?: P) => Promise<T>;
+
+export interface PollingConfigurations {
+  tabId: string;
+}
+
+export class UsePolling<T, P = void> {
+  public data: T | null = null;
+  public error: Error | null = null;
+  public loading: boolean = true;
+  private shouldPoll: boolean = false;
+  private intervalRef?: NodeJS.Timeout;
+
+  constructor(
+    private fetchFunction: FetchFunction<T, P>,
+    private interval: number = 5000,
+    private onPollingSuccess?: (data: T, configurations: PollingConfigurations) => boolean,
+    private onPollingError?: (error: Error) => boolean,
+    private configurations?: PollingConfigurations
+  ) {}
+
+  async fetchData(params?: P) {
+    this.loading = true;
+    try {
+      const result = await this.fetchFunction(params);
+      this.data = result;
+      this.loading = false;
+
+      if (this.onPollingSuccess && this.onPollingSuccess(result, this.configurations!)) {
+        this.stopPolling();
+      }
+    } catch (err) {
+      this.error = err as Error;
+      this.loading = false;
+
+      if (this.onPollingError && this.onPollingError(this.error)) {
+        this.stopPolling();
+      }
+    }
+  }
+
+  async startPolling(params?: P) {
+    this.shouldPoll = true;
```

Comment:
how come we have a boolean here vs using the interval?

Reply:
Should be lifted straight from [`use_polling.ts`](https://github.com/opensearch-project/dashboards-observability/blob/main/public/components/hooks/use_polling.ts) in `dashboards-observability`.